Orleans.Providers.MongoDB/Membership/Store/Multiple/MultipleMembershipCollection.cs

```diff
@@ -0,0 +1,199 @@
+        {
+            using (var session = await Client.StartSessionAsync())
+            {
+                try
+                {
+                    session.StartTransaction();
+
+                    var hasUpsertedTable = await tableVersionCollection.UpsertAsync(session, tableVersion, deploymentId);
+
+                    if (!hasUpsertedTable)
+                    {
+                        await session.AbortTransactionAsync();
+                        return false;
+                    }
+
+                    var hasUpsertedMember = await UpsertRowAsync(session, deploymentId, entry, etag);
+
+                    var isOkay = hasUpsertedMember && hasUpsertedTable;
+
+                    if (isOkay)
+                    {
+                        await session.CommitTransactionAsync();
+                    }
+                    else
+                    {
+                        await session.AbortTransactionAsync();
+                    }
+
+                    return isOkay;
+                }
+                catch (Exception)
+                {
+                    await session.AbortTransactionAsync();
+                    throw;
+                }
+            }
+        }
+
+        private async Task<bool> UpsertRowAsync(IClientSessionHandle session, string deploymentId, MembershipEntry entry, string etag)
+        {
+            var id = ReturnId(deploymentId, entry.SiloAddress);
+
+            var document = MongoMembershipDocument.Create(entry, deploymentId, id);
+
+            try
+            {
+                await Collection.ReplaceOneAsync(session, x => x.Id == id && x.Etag == etag, document, Upsert);
+
+                return true;
+            }
+            catch (MongoException ex)
+            {
+                if (ex.IsDuplicateKey())
+                {
+                    return false;
+                }
+                throw;
+            }
+        }
+
+        public async Task<IList<Uri>> GetGateways(string deploymentId)
+        {
+            var entries =
+                await Collection.Find(x => x.DeploymentId == deploymentId && x.Status == (int)SiloStatus.Active && x.ProxyPort > 0)
+                    .ToListAsync();
+
+            return entries.Select(x => x.ToGatewayUri()).ToList();
+        }
+
+        public async Task<MembershipTableData> ReadAll(string deploymentId)
+        {
+            using (var session = await Client.StartSessionAsync())
+            {
+                try
+                {
+                    session.StartTransaction();
+
+                    var tableVersion = tableVersionCollection.GetTableVersionAsync(deploymentId);
+
+                    var entries =
+                        Collection.Find(x => x.DeploymentId == deploymentId)
+                            .ToListAsync();
+
+                    await Task.WhenAll(tableVersion, entries);
+
+                    return ReturnMembershipTableData(entries.Result, tableVersion.Result);
```

Comment:
Should there be a `CommitTransactionAsync` call here and elsewhere?

Reply:
I don't think so, as we do not make any updates.

Reply:
That seems odd - is the transaction just hanging in that case or does it automatically abort when the session scope ends?

Reply:
You are probably right, the end session will stop it, but it is blocking. So I think it is better to commit manually.